### PR TITLE
Fix FBT001 false negative with unions and optional

### LINKED ISSUE #7487: FBT001 Boolean trap surprising false negative with `bool | None`
`FBT001` should catch boolean positional arguments in function definitions, but to my surprise it does not care about  `bool | None`, which I would expect it to do. Thinking about it, I would expect it to catch any union type where `bool` is one of the options.

In both cases, FBT003 triggers on the call to the function, but the function definition itself passes.

```python
def chaught(x: bool) -> bool:
    """FBT0001 trigger."""
    return x


def passes(x: bool | None) -> bool | None:
    """No FBT0001 trigger."""
    return x


chaught(True) #FBT003 trigger
passes(False) #FBT003 trigger
```

### PR DESCRIPTION
**Pull request by JonathanPlasse**
## Summary

- Close #7487

In the spirit of `flake8-boolean-trap`, any positional argument that can accept a boolean should raise `FBT001`.
Raise `FBT001` for all annotations that accept booleans (e.g. `Optional[bool]`, `Union[int, bool]`).

## Test Plan

Add a fixture, with an annotation using `|`, `Optional`, and `Union`, and containing a boolean.

### CODE REVIEW

**Comment by JonathanPlasse** [comment]
This could introduce a lot of new positives.
Is it possible to have a preview of rule behavior changes?

File: crates/ruff_linter/src/rules/flake8_boolean_trap/rules/boolean_type_hint_positional_argument.rs [hunk 1]
```diff
@@ -96,6 +97,40 @@ impl Violation for BooleanTypeHintPositionalArgument {
...
 }
 
+fn match_annotation_to_bool(annotation: &Expr, semantic: &SemanticModel) -> bool {
+    // check for both bool (python class) and 'bool' (string annotation)
+    let hint = match annotation {
+        Expr::Name(name) => &name.id == "bool",
+        Expr::Constant(ast::ExprConstant {
+            value: Constant::Str(ast::StringConstant { value, .. }),
+            ..
+        }) => value == "bool",
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/flake8_boolean_trap/rules/boolean_type_hint_positional_argument.rs:107]: With https://github.com/astral-sh/ruff/pull/8064 merged, you'll need to update this to use `Expr::StringLiteral` instead:

```suggestion
        Expr::StringLiteral(ast::ExprStringLiteral { value, .. }) => value == "bool",
```

**Comment by JonathanPlasse** [comment]
How should I proceed to make this behavior change only for the preview mode?

**Comment by charliermarsh** [comment]
@JonathanPlasse - I think you could change `match_annotation_to_bool` to take the preview flag as argument, and fork internally?

File: crates/ruff_linter/src/rules/flake8_boolean_trap/rules/boolean_type_hint_positional_argument.rs [hunk 2]
```diff
@@ -96,6 +97,37 @@ impl Violation for BooleanTypeHintPositionalArgument {
...
+                    false
+                }
+            } else if semantic.match_typing_expr(value, "Optional") {
+                match_annotation_to_bool(slice, semantic)
+            } else {
+                false
+            }
+        }
+        _ => false,
+    };
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_boolean_trap/rules/boolean_type_hint_positional_argument.rs:127]: @dhruvmanila - Is it possible to use `TypingTarget` for this?

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**